### PR TITLE
Fixes an issue with the boolean values in the Google Analytics Connector

### PR DESCRIPTION
**Pull request by valadas**
Right or wrong the connectors infrastructure only supports passing strings as config values. Knockout has a strange behaviour where it treats any string as true and empty strings as false. However we get back from the connector UI either "true" or "false" for some reason. This PR handles this difference of behaviour when sending and receiving the settings for boolean values.

I have tested that it saves "false" or "true" in the database and that this PR restores the module UI problem, however I am not sure then where those database values are used, so testing that could be nice :)

It also adds a confirmation on save since that part was missing compared to the Azure Connector.

### CODE REVIEW

**Comment by valadas** [comment]
@mitchelsellers @ohine I think this completes 9.4.0

File: DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs [hunk 1]
```diff
@@ -88,11 +87,14 @@ public bool HasConfig(int portalId)
...
             var portalSettings = new PortalSettings(portalId);
 
-            var trackingId = String.Empty;
-            var urlParameter = String.Empty;
-            var trackForAdmin = false;
-            var anonymizeIp = false;
-            var trackUserId = false;
+            // Important, knockout handles emptry strings as false and any other string as true
+            // so we need to pass empty strings when we mean false, however it passes us back the string "false"
+            // when saving the settins in the SaveConfig method, so we need to handle that case too
```
Comment (bdukes) [inline DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs]:
```suggestion
            // when saving the settings in the SaveConfig method, so we need to handle that case too
```

File: DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs [hunk 2]
```diff
@@ -88,11 +87,14 @@ public bool HasConfig(int portalId)
             var analyticsConfig = AnalyticsConfiguration.GetConfig("GoogleAnalytics");
             var portalSettings = new PortalSettings(portalId);
 
-            var trackingId = String.Empty;
-            var urlParameter = String.Empty;
-            var trackForAdmin = false;
-            var anonymizeIp = false;
-            var trackUserId = false;
+            // Important, knockout handles emptry strings as false and any other string as true
```
Comment (bdukes) [inline DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs]:
```suggestion
            // Important, knockout handles empty strings as false and any other string as true
```

File: DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs [hunk 3]
```diff
@@ -108,46 +110,52 @@ public bool HasConfig(int portalId)
...
-                { "TrackAdministrators", trackForAdmin.ToString()},
-                { "AnonymizeIp", anonymizeIp.ToString()},
-                { "TrackUserId", trackUserId.ToString()},
-                { "DataConsent", portalSettings.DataConsentActive.ToString()},
-                { "isDeactivating", false.ToString()}
+                { "TrackAdministrators", trackForAdmin},
+                { "AnonymizeIp", anonymizeIp},
+                { "TrackUserId", trackUserId},
+                { "DataConsent", portalSettings.DataConsentActive ? "true" : ""},
+                { "isDeactivating", ""}
```
Comment (bdukes) [inline DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs]:
```suggestion
                { "isDeactivating", HandleCustomBoolean(false) }
```

File: DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs [hunk 4]
```diff
@@ -108,46 +110,52 @@ public bool HasConfig(int portalId)
...
                 { "UrlParameter", urlParameter},
-                { "TrackAdministrators", trackForAdmin.ToString()},
-                { "AnonymizeIp", anonymizeIp.ToString()},
-                { "TrackUserId", trackUserId.ToString()},
-                { "DataConsent", portalSettings.DataConsentActive.ToString()},
-                { "isDeactivating", false.ToString()}
+                { "TrackAdministrators", trackForAdmin},
+                { "AnonymizeIp", anonymizeIp},
+                { "TrackUserId", trackUserId},
+                { "DataConsent", portalSettings.DataConsentActive ? "true" : ""},
```
Comment (bdukes) [inline DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs]:
```suggestion
                { "DataConsent", HandleCustomBoolean(portalSettings.DataConsentActive) },
```

File: DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs [hunk 5]
```diff
@@ -165,28 +173,27 @@ public bool SaveConfig(int portalId, IDictionary<string, string> values, ref boo
...
                 }
                 else
                 {
-                    trackingID = values["TrackingID"] != null ? values["TrackingID"].ToString().Trim() : String.Empty;
-                    urlParameter = values["UrlParameter"] != null ? values["UrlParameter"].ToString().Trim() : String.Empty;
-
-                    bool.TryParse(values["TrackAdministrators"].ToLowerInvariant(), out trackForAdmin);
-                    bool.TryParse(values["AnonymizeIp"].ToLowerInvariant(), out anonymizeIp);
-                    bool.TryParse(values["TrackUserId"].ToLowerInvariant(), out trackUserId);
+                    trackingID = values["TrackingID"] != null ? values["TrackingID"].ToLowerInvariant().Trim() : string.Empty;
```
Comment (bdukes) [inline DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs]: should tracking ID and URL parameter be lower-cased?

File: DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs [hunk 6]
```diff
@@ -108,46 +110,52 @@ public bool HasConfig(int portalId)
...
 
+        /// <summary>
+        /// Handles custom conversion from "true" => "true"
+        /// Anything else to "" to support the strange knockout handling of string as booleans
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        private string HandleCustomBoolean(string value)
+        {
+            if (value.ToLowerInvariant().Trim() == "true")
```
Comment (bdukes) [inline DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs]:
```suggestion
            if (value.Trim().Equals("true", StringComparison.OrdinalIgnoreCase))
```

**Comment by mitchelsellers** [comment]
@valadas if you can address the above items I can get a review in.  I tried to commit a few of the changes, but it looks like your branch protection doesn't allow edits from us.

File: DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs (hunk 5 above)
Comment (valadas) [reply DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs]: Nope, the object is lower case but the key is Upper Case

**Comment by valadas** [comment]
@mitchelsellers @bdukes , changes done and had to modify one suggestion because HandleCustomBoolean takes a string, not a bool. Should be good to go now.

File: DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs (hunk 5 above)
Comment (bdukes) [reply DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs]: @valadas I'm asking about the `ToLowerInvariant()` call on the value, that shouldn't be there, should it?
```suggestion
                    trackingID = values["TrackingID"]?.Trim() ?? string.Empty;
```

File: DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs [hunk 7]
```diff
@@ -165,28 +173,27 @@ public bool SaveConfig(int portalId, IDictionary<string, string> values, ref boo
...
                 else
                 {
-                    trackingID = values["TrackingID"] != null ? values["TrackingID"].ToString().Trim() : String.Empty;
-                    urlParameter = values["UrlParameter"] != null ? values["UrlParameter"].ToString().Trim() : String.Empty;
-
-                    bool.TryParse(values["TrackAdministrators"].ToLowerInvariant(), out trackForAdmin);
-                    bool.TryParse(values["AnonymizeIp"].ToLowerInvariant(), out anonymizeIp);
-                    bool.TryParse(values["TrackUserId"].ToLowerInvariant(), out trackUserId);
+                    trackingID = values["TrackingID"] != null ? values["TrackingID"].ToLowerInvariant().Trim() : string.Empty;
+                    urlParameter = values["UrlParameter"] != null ? values["UrlParameter"].ToLowerInvariant().Trim() : string.Empty;
```
Comment (bdukes) [inline DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs]:
```suggestion
                    urlParameter = values["UrlParameter"]?.Trim() ?? string.Empty;
```

Comment (valadas) [reply DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs]: Any reason for removing the .ToLowerInvariant or is that a mistake?

Comment (bdukes) [reply DNN Platform/Connectors/GoogleAnalytics/GoogleAnalyticsConnector.cs]: It seemed that `ToLowerInvariant` was added to make sure that `True` was turned into `true` for the Boolean settings, but I don't see any reason to make the values of the two text settings always be lower case.  I assumed it was just copied from the Boolean case without realizing.

**Pull request merged**